Use linecache.lazycache to cache generated code.

davides opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63453

Summary:
Instead of patching linecache.getlines, use linecache.lazycache and
parts of the loader protocol described in PEP-302

Test Plan:
python3 test/test_fx.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30388176](https://our.internmc.facebook.com/intern/diff/D30388176)

davides committed:
Use linecache.lazycache to cache generated code.

Summary:
Instead of patching linecache.getlines, use linecache.lazycache and
parts of the loader protocol described in PEP-302

Test Plan:
python3 test/test_fx.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

davides committed:
Update on "Use linecache.lazycache to cache generated code."


Summary:
Instead of patching linecache.getlines, use linecache.lazycache and
parts of the loader protocol described in PEP-302

Test Plan:
python3 test/test_fx.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

davides commented:
@davides has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30388176).

suo commented on torch/fx/graph_module.py:
```diff
@@ -15,28 +15,60 @@
+    def get_source(self, module_name) -> str:
+        if module_name in self.eval_cache:
+            return self.eval_cache[module_name]
+        return None
```
this is causing mypy to fail—looks like PEP302 implies it's an optional[str]

davides replied:
Good catch, updated.

suo reviewed (changes requested):
Needs a small fix but overall looks pretty good!

davides committed:
Update on "Use linecache.lazycache to cache generated code."


Summary:
Instead of patching linecache.getlines, use linecache.lazycache and
parts of the loader protocol described in PEP-302

Test Plan:
python3 test/test_fx.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30388176](https://our.internmc.facebook.com/intern/diff/D30388176)

[ghstack-poisoned]

suo reviewed (approved):
lgtm, provided tests pass!

suo commented:
oops, I accepted a little too early, see comment on https://github.com/pytorch/pytorch/pull/63497

davides committed:
Update on "Use linecache.lazycache to cache generated code."


Summary:
Instead of patching linecache.getlines, use linecache.lazycache and
parts of the loader protocol described in PEP-302

Test Plan:
python3 test/test_fx.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30388176](https://our.internmc.facebook.com/intern/diff/D30388176)

[ghstack-poisoned]

davides commented:
@davides has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30388176).

davides committed:
Update on "Use linecache.lazycache to cache generated code."


Summary:
Instead of patching linecache.getlines, use linecache.lazycache and
parts of the loader protocol described in PEP-302

Test Plan:
python3 test/test_fx.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30388176](https://our.internmc.facebook.com/intern/diff/D30388176)

[ghstack-poisoned]

davides commented:
@davides has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30388176).
